Protobuf specification rework (#131)
disneystreaming/alloy: docs/serialisation/protobuf.md

## Protobuf serialisation

When protocols use protobuf as a serialisation, alloy also proposes a set of semantics for how smithy shapes translate to protobuf-specific concepts. In this document, we describe these semantics by explaining how the smithy code should translate to proto code representing the equivalent data.

`alloy` defines a number of traits that are aimed at capturing various semantics relative to [https://protobuf.dev/overview/](protobuf) serialisation that are not part of the smithy standard library.

For full documentation on what each of these traits does, see the smithy specification [here](modules/core/resources/META-INF/smithy/proto/proto.smithy).

Note that for convenience, `alloy` provides a module containing protobuf definitions that used downstream to ensure that the semantics described in this document are respected.

```
com.disneystreaming.alloy:alloy-protocol:x.y.z
```

### Validation

> `alloy` comes with validator that verifies the abidance of shapes to the rules described below. Note that these validators are protocol-specific, and

Reviewer: ```suggestion
`alloy` comes with validators that verify the abidance of shapes to the rules described below. Note that these validators are protocol-specific, and
```

| integer, byte, short | SIGNED        | alloy.protobuf.SInt32Value      |
| integer, byte, short | UNSIGNED      | google.protobuf.UInt32Value     |
| long                 | N/A           | google.protobuf.SInt64Value     |
| long                 | FIXED         | alloy.protobuf.Fixed64Value     |
| long                 | FIXED_SIGNED  | alloy.protobuf.SFixed64Value    |
| long                 | SIGNED        | alloy.protobuf.SInt64Value      |
| long                 | UNSIGNED      | google.protobuf.UInt64Value     |

#### alloy.proto#protoNumType

Integer and Long shapes can be annotated with the `@alloy.proto#protoNumType` in order to signal what encoding should be used during protobuf serialisation.

- SIGNED
- UNSIGNED
- FIXED
- FIXED_SIGNED

See [here](https://protobuf.dev/programming-guides/proto3/#scalar) for documentation about these encodings.

#### UUIDs

By default, string shapes annotated with `@uuidFormat` are serialised as protobuf strings. However, a `alloy.proto#protoCompactUUID` trait is provided, which signals that the serialised form should be a message containing two int64 values :

Smithy:

```smithy
use alloy#uuidFormat
use alloy.proto#protoCompactUUID

> @protoCompactUUID

Reviewer: smithy translate did that for shapes with the type:`alloy#UUID`

https://github.com/disneystreaming/smithy-translate/blob/main/modules/proto-examples/protobuf/demo/definitions.proto#L24

Author: I'm aware. We shouldn't use shapeIds to discriminate behaviour, it's pretty unsound. Using traits is much better. 

I guess your point is that we could apply that trait to the `alloy#UUID` idea to retain the current behaviour.

Reviewer: I did not know you were aware, I agree with you that the trait is a better approach now that I know better!
But as jeff points out, I don't think introducing a dependency between the two artifacts is desired so I would just leave it like that

@uuidFormat
string MyUUID

structure Foo {
  uuid : alloy#UUID
}
```

Proto:

```proto
message MyUUID {
  int64 upper_bits = 1;
  int64 lower_bits = 2;
}

message Foo {
  uuid: MyUUID
}
```

### Documents

Documents should be serialised using a protobuf message equivalent to the one below :

```proto
message DNull {
}

message DArray {
  repeated Document values = 1;
}

message DObject {
  map<string, Document> values = 1;
}

message Document {
  oneof value {
    DNull dNull = 1;
    bool dBoolean = 2;
    double dNumber = 3;
    string dString = 4;
    DArray dArray = 5;
    DObject dObject = 6;
  }
}
```

For convenience, alloy provides this proto definition in a jar publised to maven-central `alloy-protobuf`

### Aggregate Types

Structure, unions and (string) enumerations follow the rule that unless specified otherwise, the "proto indexes" (or "field numbers") assigned

> to their members should be monotonically increase. The first member receives the index `1` in the case of structures and unions, whilst it receives the index `0` in the case of enumerations.

Reviewer 2: For structures at least, we don't enforce that the first index is 1 or that they increase. As far as I can tell, the only validation is that there are no duplicate indices. I'm not sure about unions

Author: @lewisjkl, in this paragraph I'm talking about what happens in the absence of `protoIndex`. Granted it's poorly phrased, I'll work on it.

Author: [b3faf4f](https://github.com/disneystreaming/alloy/pull/131/commits/b3faf4f9dcc5a2228d72385c8641334eef4bbc32)

Reviewer 2: Ah I see, and yeah that revised version sounds good to me. Thanks

}
```

##### Union members targeting collections

Protobuf doesn't allow `oneof` members to have `repeated` or `map` fields. As a result, a smithy union with a members targeting a collection shapes MUST
either have the `@protoWrapped` trait or target a collection shape have the `@protoWrapped` trait.

##### Inlined unions (`alloy.proto#protoInlinedOneOf`)

> The `alloy.proto#protoInlinedOneOf` trait can be used to inline the corresponding `oneof` in a protobuf message. A union with this trait MUST be used referenced exactly once, by a structure member.

Reviewer: ```suggestion
The `alloy.proto#protoInlinedOneOf` trait can be used to inline the corresponding `oneof` in a protobuf message. A union with this trait MUST be used exactly once, by a structure member.
```

}
```

> has the following meaning in proto semantics

Reviewer: ```suggestion
has the following meaning in protobuf semantics
```

##### protoIndex for enumerations

> Members of closed enumerations (whether string or int) can be annotated by `alloy.proto#protoIndex` in smithy to customise the corresponding proto index that should be used during serialisation. An additional constraint is that when users elect to specify `alloy.proto#protoIndex`, they are required to assign the `0` value to one of the enumeration members, as it is a requirement on the protobuf side.

Reviewer: ```suggestion
Members of closed enumerations (whether string or int) can be annotated by `alloy.proto#protoIndex` in smithy to customise the corresponding proto index that should be used during serialisation. An additional constraint is that when users elect to specify `alloy.proto#protoIndex`, they are required to assign the `0` value to one of the enumeration members, as it is a requirement for protobuf.
```
